File: code_source_upscale/upscale.js
/**
 * Upscale minimal : une entrée -> une sortie upscalée (x4).
 * Utilise le moteur upscayl-bin (Real-ESRGAN / NCNN Vulkan) fourni dans bin/.
 *
 * Usage :
 *     node upscale.js photo.jpg photo_upscaled.png
 */
const { execFile } = require("node:child_process");
const path = require("node:path");

const BINAIRE = path.join(__dirname, "bin", "upscayl-bin.exe");
const MODELES = path.join(__dirname, "models");

/**
 * @param {string} entree  chemin de l'image source (jpg/png/webp)
 * @param {string} sortie  chemin de l'image upscalée à produire
 * @param {string} [modele] "upscayl-standard-4x" (qualité) ou "upscayl-lite-4x" (rapide)
 * @param {number} [echelle] 2, 3 ou 4
 * @returns {Promise<void>}
 */
function upscale(entree, sortie, modele = "upscayl-standard-4x", echelle = 4) {
  return new Promise((resolve, reject) => {
    execFile(
      BINAIRE,
      ["-i", entree, "-o", sortie, "-m", MODELES, "-n", modele, "-s", String(echelle)],
      (err, _stdout, stderr) => (err ? reject(new Error(`Echec de l'upscale : ${stderr}`)) : resolve())
    );
  });
}

module.exports = { upscale };

if (require.main === module) {
  const [entree, sortie, modele, echelle] = process.argv.slice(2);
  if (!entree || !sortie) {
    console.error("Usage : node upscale.js <entree> <sortie> [modele] [echelle]");
    process.exit(1);
  }
  upscale(entree, sortie, modele, echelle ? Number(echelle) : undefined)
    .then(() => console.log(`OK : ${sortie}`))
    .catch((e) => { console.error(e.message); process.exit(1); });
}

File: code_source_upscale/upscale.py
"""
Upscale minimal : une entrée -> une sortie upscalée (x4).
Utilise le moteur upscayl-bin (Real-ESRGAN / NCNN Vulkan) fourni dans bin/.

Usage :
    python upscale.py photo.jpg photo_upscaled.png
"""
import subprocess
import sys
from pathlib import Path

RACINE = Path(__file__).parent
BINAIRE = RACINE / "bin" / "upscayl-bin.exe"
MODELES = RACINE / "models"


def upscale(entree: str, sortie: str, modele: str = "upscayl-standard-4x", echelle: int = 4) -> None:
    """Upscale l'image `entree` et écrit le résultat dans `sortie`.

    modele  : "upscayl-standard-4x" (qualité) ou "upscayl-lite-4x" (rapide)
    echelle : 2, 3 ou 4
    """
    resultat = subprocess.run(
        [
            str(BINAIRE),
            "-i", str(entree),
            "-o", str(sortie),
            "-m", str(MODELES),
            "-n", modele,
            "-s", str(echelle),
        ],
        capture_output=True,
        text=True,
        encoding="utf-8",
        errors="replace",
    )
    if resultat.returncode != 0:
        raise RuntimeError(f"Echec de l'upscale : {resultat.stderr}")


if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Usage : python upscale.py <entree> <sortie> [modele] [echelle]")
        sys.exit(1)
    upscale(
        sys.argv[1],
        sys.argv[2],
        sys.argv[3] if len(sys.argv) > 3 else "upscayl-standard-4x",
        int(sys.argv[4]) if len(sys.argv) > 4 else 4,
    )
    print(f"OK : {sys.argv[2]}")
